### ISSUE TITLE
Core requirements for router to router links

### ISSUE THREAD

**Issue by om26er** [wamp-proto/wamp-proto#399]
Crossbar implements router to router links. Its a WAMP based mechanism of connecting multiple routers to each other and effectively creating a mesh of routers. It can also be useful for edge-cloud scenarios. Ultimately allowing different type of WAMP routers to connect to each other.

In practice the router to router links code could just be a pluggable WAMP component that can be used with any router. The only change that WAMP routers would need to do is in their broker and dealer code to allow to disclose the identity of the actual caller instead of the relay.

Imagine a group of three routers
* Router 1 connected to router 2
* Router 2 connected to router 3

For an event to reach from router 1 to router 3, it would need to go through router 2 BUT which caller will be disclosed when the call actually reaches the callee connected to router 3 ? Currently, the callee would be the session between router 2 and router 3, however the actual caller is the client connected to router 1.

To deal with the above situation Crossbar uses `forward_for`, which is a list that contains the sessions the message has hoped through. https://github.com/crossbario/crossbar/blob/cce29666fd0c173a41ea6e33c7e19b8a0427458b/crossbar/router/dealer.py#L953

----
My proposal is that we should work to define the core of router to router links in the first phase

**Comment by om26er** [comment]
Inspired by Crossbar's router to router links code, I have created a small project that extends Nexus (another WAMP router) to implement basic router to router links. If needed we can enhance that project so that all interested WAMP routers could use that code has a way to ensure they can work with rlinks
https://github.com/om26er/nexlinks

**Comment by oberstet** [comment]
Interesting;) yeah, I'd be interested in joining a discussion! the whole thing, and the code in Crossbar.io are already at a stage where we can actually experiment, adjust, iterate .. also rgd the technical spec and details. However, it is _not_ yet finished, but experimental. Eg. in crossbar, besides `forward_for`, we will use `authextra` for extending WAMP-cryptosign for the authentication of rlinks - to enable cross-implementation and cross-operator rlink.

> we should work to define the core of router to router links in the first phase

what do you mean with "core"? the requirements or the technical design or the detailed spec?

**Comment by om26er** [comment]
> what do you mean with "core"? the requirements or the technical design or the detailed spec?

I basically mean that we should work on a draft that would document what changes the routing code would need to do to support router to router links. `forward_for` is a prime example as that's a necessity even for the simplest form of router to router link.

That way we can essentially make any router that implements WAMP session/registration/subscription meta APIs into an "edge" router, potentially enabling new ways WAMP would be used.

**Comment by om26er** [comment]
`forward_for` or something similar is important as when a call is routed through multiple routers, we need to "disclose" the correct caller when the call reaches the callee. In my PR to Nexus, I am trying to get this basic functionality landed so that the correct caller/publisher could be disclosed. https://github.com/gammazero/nexus/pull/268

**Comment by oberstet** [comment]
> forward_for or something similar is important as when a call is routed through multiple routers, we need to "disclose" the correct caller when the call reaches the callee.

I think you are misunderstanding the use of `forward_for` .. being able to disclose forward_for[0] to the actual target is just one use of it. the more important aspect is the source routing through node meshes.

> In my PR to Nexus, I am trying to get this basic functionality landed 

what functionality? there is no spec, so how would this PR be evaluated (against what?)

**Comment by oberstet** [comment]
but trying to be more constructive: I would say a PR for only the `forward_for` field .. is just shooting in the dark currently. one change that is upcoming (as hinted above) is the move to "global IDs" .. otherwise the meaning of the IDs in forward_for is bogus. so discussing these aspects would be very helpful in my eyes

IOW: this code will likely change:

https://github.com/crossbario/autobahn-python/blob/586548222916b4f2a144b13959e1458186a4d923/autobahn/wamp/message.py#L1480

**Comment by om26er** [comment]
> I think you are misunderstanding the use of `forward_for` .. being able to disclose forward_for[0] to the actual target is just one use of it. the more important aspect is the source routing through node meshes.

I am aware of that, in my initial experiment of edge-cloud, I only needed a way to disclose the caller/publisher and took inspiration from Crossbar. I do understand that its not part of the spec but it will unblock me to do take forward my experiment. (Quite likely the PR won't event land).

> but trying to be more constructive: I would say a PR for only the `forward_for` field .. is just shooting in the dark currently. one change that is upcoming (as hinted above) is the move to "global IDs" .. otherwise the meaning of the IDs in forward_for is bogus. so discussing these aspects would be very helpful in my eyes

Thanks for the feedback. I am looking forward to seeing "global IDs" in Crossbar/Autobahn, I am sure I will have feedback on it
